Framework/HistogramData/inc/MantidHistogramData/VectorOf.h

```diff
@@ -0,0 +1,158 @@
+  VectorOf(size_t count, const double &value) {
+    m_data = Kernel::make_cow<CowType>(count, value);
+  }
+  /// Construct stored object of length count.
+  explicit VectorOf(size_t count) { m_data = Kernel::make_cow<CowType>(count); }
+  /// Construct stored object with the contents of the initializer list init.
+  VectorOf(std::initializer_list<double> init) {
+    m_data = Kernel::make_cow<CowType>(init);
+  }
+  /// Copy constructor. Lightweight, stored object will be shared.
+  VectorOf(const VectorOf &) = default;
+  /// Move constructor.
+  VectorOf(VectorOf &&) = default;
+  // Note the lvalue reference qualifier for all assignment operators. This
+  // prevents mistakes in client code, assigning to an rvalue, such as
+  // histogram.getBinEdges() = { 0.1, 0.2 };
+  /// Copy assignment. Lightweight, stored object will be shared.
+  VectorOf &operator=(const VectorOf &)& = default;
+  /// Move assignment.
+  VectorOf &operator=(VectorOf &&)& = default;
+
+  /// Assigns the stored object with the contents of the initializer list init.
+  VectorOf &operator=(std::initializer_list<double> ilist) & {
+    m_data = Kernel::make_cow<CowType>(ilist);
+    return *this;
+  }
+  /// Constructs the stored object with the contents of the range [first, last).
+  template <class InputIt>
+  VectorOf(InputIt first, InputIt last)
+      : m_data(Kernel::make_cow<CowType>(first, last)) {}
+
+  /// Copy construct from cow_ptr. Lightweight, stored object will be shared.
+  explicit VectorOf(const Kernel::cow_ptr<CowType> &other) : m_data(other) {}
+  /// Copy construct from shared_ptr. Lightweight, stored object will be shared.
+  explicit VectorOf(const boost::shared_ptr<CowType> &other) : m_data(other) {}
+  /// Copy construct stored object from data.
+  explicit VectorOf(const CowType &data)
+      : m_data(Kernel::make_cow<CowType>(data)) {}
+  /// Move construct stored object from data.
+  explicit VectorOf(CowType &&data)
+      : m_data(Kernel::make_cow<CowType>(std::move(data))) {}
+
+  /// Copy assignment from cow_ptr. Lightweight, stored object will be shared.
+  VectorOf &operator=(const Kernel::cow_ptr<CowType> &other) & {
+    m_data = other;
+    return *this;
+  }
+  /// Copy assignment from shared_ptr. Lightweight, stored object will be
+  /// shared.
+  VectorOf &operator=(const boost::shared_ptr<CowType> &other) & {
+    m_data = other;
+    return *this;
+  }
+  /// Copy assignment to stored object from data.
+  VectorOf &operator=(const CowType &data) & {
+    if (!m_data || (&(*m_data) != &data))
+      m_data = Kernel::make_cow<CowType>(data);
+    return *this;
+  }
+  /// Move assignment to stored object from data.
+  VectorOf &operator=(CowType &&data) & {
+    if (!m_data || (&(*m_data) != &data))
+      m_data = Kernel::make_cow<CowType>(std::move(data));
+    return *this;
+  }
+
+  /// Checks if *this stores a non-null pointer.
+  explicit operator bool() const { return m_data.operator bool(); }
```

Comment:
Is this the right way to call the overloaded function operator on `m_data`? I haven't seen it in this form before?